// tih-web/src/lib/demoData.ts
import type { TalentProfile, Tag, Evaluation, Activity, DashboardStats } from '../types/database'

// ─── Demo Organization & Users ─────────────────────────────────────────────────
export const DEMO_ORG_ID = 'demo-org-001'
export const DEMO_USER_ID = 'demo-user-001'

export const demoTags: Tag[] = [
  { id: 't1', organization_id: DEMO_ORG_ID, name: 'AI/ML', color: '#8b5cf6', created_at: '2024-01-01T00:00:00Z' },
  { id: 't2', organization_id: DEMO_ORG_ID, name: 'Python', color: '#3b82f6', created_at: '2024-01-01T00:00:00Z' },
  { id: 't3', organization_id: DEMO_ORG_ID, name: 'React', color: '#06b6d4', created_at: '2024-01-01T00:00:00Z' },
  { id: 't4', organization_id: DEMO_ORG_ID, name: 'Remote', color: '#10b981', created_at: '2024-01-01T00:00:00Z' },
  { id: 't5', organization_id: DEMO_ORG_ID, name: 'Enterprise', color: '#f59e0b', created_at: '2024-01-01T00:00:00Z' },
  { id: 't6', organization_id: DEMO_ORG_ID, name: 'DevOps', color: '#ef4444', created_at: '2024-01-01T00:00:00Z' },
  { id: 't7', organization_id: DEMO_ORG_ID, name: 'Leadership', color: '#ec4899', created_at: '2024-01-01T00:00:00Z' },
  { id: 't8', organization_id: DEMO_ORG_ID, name: 'Data Science', color: '#14b8a6', created_at: '2024-01-01T00:00:00Z' },
]

export const demoProfiles: TalentProfile[] = [
  {
    id: 'p1', organization_id: DEMO_ORG_ID, full_name: 'Priya Sharma',
    email: 'priya.sharma@techcorp.com', phone: '+91-9876543210',
    talent_type: 'Trainer', source: 'LinkedIn', status: 'Shortlisted',
    organization: 'TechCorp India', designation: 'Senior Trainer',
    years_experience: 8, location: 'Chennai', availability: 'Immediate',
    expected_compensation: 120000, primary_skills: ['Python', 'Machine Learning', 'Deep Learning'],
    secondary_skills: ['TensorFlow', 'Keras', 'Data Visualization'],
    certifications: ['AWS ML Specialty', 'Google Data Engineer'],
    domains: ['EdTech', 'FinTech'], linkedin_url: 'https://linkedin.com/in/priyasharma',
    overall_rating: 4.8, is_shortlisted: true, is_favorite: true, is_active: true,
    notes: 'Exceptional trainer with proven track record in enterprise ML deployments.',
    created_at: '2024-03-15T10:30:00Z', updated_at: '2024-05-20T14:00:00Z',
    tags: [demoTags[0], demoTags[1], demoTags[7]],
  },
  {
    id: 'p2', organization_id: DEMO_ORG_ID, full_name: 'Rahul Verma',
    email: 'rahul.verma@freelance.dev', phone: '+91-9123456780',
    talent_type: 'Consultant', source: 'Referral', status: 'Under Review',
    organization: 'Self-Employed', designation: 'Cloud Architect',
    years_experience: 12, location: 'Bangalore', availability: '2 Weeks',
    expected_compensation: 200000, primary_skills: ['AWS', 'Azure', 'Kubernetes'],
    secondary_skills: ['Terraform', 'Ansible', 'Docker'],
    certifications: ['AWS Solutions Architect Pro', 'CKA', 'Azure Solutions Architect'],
    domains: ['Cloud', 'DevOps', 'Enterprise IT'], linkedin_url: 'https://linkedin.com/in/rahulverma',
    overall_rating: 4.5, is_shortlisted: false, is_favorite: false, is_active: true,
    created_at: '2024-04-02T09:00:00Z', updated_at: '2024-05-18T11:30:00Z',
    tags: [demoTags[5], demoTags[4]],
  },
  {
    id: 'p3', organization_id: DEMO_ORG_ID, full_name: 'Meera Krishnan',
    email: 'meera.k@design.studio', phone: '+91-9988776655',
    talent_type: 'Freelancer', source: 'WhatsApp', status: 'New',
    organization: 'Design Studio', designation: 'UX Lead',
    years_experience: 6, location: 'Mumbai', availability: 'Immediate',
    expected_compensation: 90000, primary_skills: ['Figma', 'UI/UX Design', 'Design Systems'],
    secondary_skills: ['React', 'CSS', 'Prototyping'],
    certifications: ['Google UX Design Certificate'],
    domains: ['Product Design', 'Consumer Apps'], portfolio_url: 'https://meera.design',
    overall_rating: 4.2, is_shortlisted: false, is_favorite: true, is_active: true,
    created_at: '2024-05-01T08:00:00Z', updated_at: '2024-05-10T10:00:00Z',
    tags: [demoTags[2], demoTags[3]],
  },
  {
    id: 'p4', organization_id: DEMO_ORG_ID, full_name: 'Arjun Mehta',
    email: 'arjun.mehta@corp.io', phone: '+91-9871234567',
    talent_type: 'Employee', source: 'Job Portal', status: 'Approved',
    organization: 'InnoSoft Ltd', designation: 'Full Stack Developer',
    years_experience: 5, location: 'Hyderabad', availability: '1 Month',
    expected_compensation: 110000, primary_skills: ['React', 'Node.js', 'TypeScript'],
    secondary_skills: ['PostgreSQL', 'Redis', 'GraphQL'],
    certifications: ['AWS Developer Associate'],
    domains: ['SaaS', 'Enterprise Software'],
    overall_rating: 4.0, is_shortlisted: true, is_favorite: false, is_active: true,
    created_at: '2024-02-20T12:00:00Z', updated_at: '2024-04-30T16:00:00Z',
    tags: [demoTags[2], demoTags[4]],
  },
  {
    id: 'p5', organization_id: DEMO_ORG_ID, full_name: 'Divya Nair',
    email: 'divya.nair@mentor.io', phone: '+91-9765432109',
    talent_type: 'Mentor', source: 'LinkedIn', status: 'Shortlisted',
    organization: 'MentorConnect', designation: 'Leadership Coach',
    years_experience: 15, location: 'Pune', availability: 'Flexible',
    expected_compensation: 180000, primary_skills: ['Leadership', 'Coaching', 'Strategy'],
    secondary_skills: ['OKRs', 'Agile', 'Change Management'],
    certifications: ['ICF PCC Certification', 'Executive Coach'],
    domains: ['Leadership Development', 'Organizational Culture'],
    overall_rating: 4.9, is_shortlisted: true, is_favorite: true, is_active: true,
    notes: 'Outstanding mentor with C-suite coaching experience across 3 continents.',
    created_at: '2024-01-15T07:00:00Z', updated_at: '2024-05-22T09:00:00Z',
    tags: [demoTags[6], demoTags[4]],
  },
  {
    id: 'p6', organization_id: DEMO_ORG_ID, full_name: 'Karthik Rajan',
    email: 'karthik.r@datascience.in', phone: '+91-9654321098',
    talent_type: 'Consultant', source: 'Email', status: 'New',
    organization: 'DataInsight Consulting', designation: 'Data Scientist',
    years_experience: 4, location: 'Chennai', availability: '3 Weeks',
    expected_compensation: 95000, primary_skills: ['Python', 'R', 'Statistics'],
    secondary_skills: ['Tableau', 'Power BI', 'SQL'],
    certifications: ['Coursera Data Science Specialization'],
    domains: ['Analytics', 'BI', 'Retail'],
    overall_rating: 3.8, is_shortlisted: false, is_favorite: false, is_active: true,
    created_at: '2024-05-08T11:00:00Z', updated_at: '2024-05-08T11:00:00Z',
    tags: [demoTags[1], demoTags[7]],
  },
  {
    id: 'p7', organization_id: DEMO_ORG_ID, full_name: 'Sneha Patel',
    email: 'sneha.patel@speakersbureau.com', phone: '+91-9543210987',
    talent_type: 'Speaker', source: 'Referral', status: 'Approved',
    organization: 'Speakers Bureau India', designation: 'Public Speaker & Author',
    years_experience: 10, location: 'Delhi', availability: 'Flexible',
    expected_compensation: 150000, primary_skills: ['Public Speaking', 'Innovation', 'Digital Transformation'],
    secondary_skills: ['Workshop Facilitation', 'Writing', 'Storytelling'],
    certifications: ['TEDx Speaker', 'NSA Member'],
    domains: ['Digital Transformation', 'Innovation', 'Leadership'],
    portfolio_url: 'https://snehapatel.com', linkedin_url: 'https://linkedin.com/in/snehapatel',
    overall_rating: 4.6, is_shortlisted: true, is_favorite: false, is_active: true,
    created_at: '2024-03-01T10:00:00Z', updated_at: '2024-05-15T13:00:00Z',
    tags: [demoTags[6], demoTags[4]],
  },
  {
    id: 'p8', organization_id: DEMO_ORG_ID, full_name: 'Vikram Singh',
    email: 'vikram.singh@contractor.dev', phone: '+91-9432109876',
    talent_type: 'Contractor', source: 'Manual', status: 'Rejected',
    organization: 'TechHire Solutions', designation: 'DevOps Engineer',
    years_experience: 3, location: 'Jaipur', availability: 'Immediate',
    expected_compensation: 75000, primary_skills: ['Docker', 'CI/CD', 'Jenkins'],
    secondary_skills: ['Python', 'Bash', 'Monitoring'],
    certifications: [],
    domains: ['DevOps', 'Automation'],
    overall_rating: 2.9, is_shortlisted: false, is_favorite: false, is_active: true,
    notes: 'Not the right fit for the current requirement. Keep for future reference.',
    created_at: '2024-04-20T14:00:00Z', updated_at: '2024-04-25T16:00:00Z',
    tags: [demoTags[5]],
  },
  {
    id: 'p9', organization_id: DEMO_ORG_ID, full_name: 'Ananya Gupta',
    email: 'ananya.g@airesearch.org', phone: '+91-9321098765',
    talent_type: 'Trainer', source: 'WhatsApp', status: 'Under Review',
    organization: 'AI Research Labs', designation: 'NLP Specialist',
    years_experience: 7, location: 'Bangalore', availability: '2 Months',
    expected_compensation: 140000, primary_skills: ['NLP', 'BERT', 'LLMs'],
    secondary_skills: ['Python', 'PyTorch', 'Hugging Face'],
    certifications: ['DeepLearning.AI NLP Specialization'],
    domains: ['AI Research', 'EdTech'],
    overall_rating: 4.3, is_shortlisted: false, is_favorite: true, is_active: true,
    created_at: '2024-05-12T09:00:00Z', updated_at: '2024-05-19T11:00:00Z',
    tags: [demoTags[0], demoTags[1]],
  },
  {
    id: 'p10', organization_id: DEMO_ORG_ID, full_name: 'Rohan Kapoor',
    email: 'rohan.kapoor@fintech.co', phone: '+91-9210987654',
    talent_type: 'Employee', source: 'LinkedIn', status: 'Engaged',
    organization: 'FinTech Innovations', designation: 'Product Manager',
    years_experience: 9, location: 'Mumbai', availability: 'Engaged',
    expected_compensation: 160000, primary_skills: ['Product Management', 'Agile', 'Roadmapping'],
    secondary_skills: ['SQL', 'Analytics', 'Stakeholder Management'],
    certifications: ['PMP', 'Certified Scrum Product Owner'],
    domains: ['FinTech', 'B2B SaaS'],
    linkedin_url: 'https://linkedin.com/in/rohankapoor',
    overall_rating: 4.7, is_shortlisted: false, is_favorite: false, is_active: true,
    created_at: '2024-01-05T08:00:00Z', updated_at: '2024-03-10T12:00:00Z',
    tags: [demoTags[4], demoTags[6]],
  },
]

export const demoEvaluations: Evaluation[] = [
  {
    id: 'e1', organization_id: DEMO_ORG_ID, talent_id: 'p1', evaluator_id: DEMO_USER_ID,
    metrics: { Communication: 5, Expertise: 5, Professionalism: 4, Problem_Solving: 5 },
    overall_score: 4.75, feedback: 'Exceptional trainer. Clear communication and deep expertise.',
    created_at: '2024-05-18T10:00:00Z',
  },
  {
    id: 'e2', organization_id: DEMO_ORG_ID, talent_id: 'p5', evaluator_id: DEMO_USER_ID,
    metrics: { Leadership: 5, Coaching: 5, Strategy: 5, Communication: 4 },
    overall_score: 4.75, feedback: 'Outstanding leadership coach. C-suite experience is a huge plus.',
    created_at: '2024-05-20T14:00:00Z',
  },
]

export const demoActivities: Activity[] = [
  {
    id: 'a1', organization_id: DEMO_ORG_ID, talent_id: 'p1',
    action: 'STATUS_CHANGED', description: 'Status changed from Under Review to Shortlisted',
    created_by: DEMO_USER_ID, created_at: '2024-05-20T14:00:00Z',
    talent: { id: 'p1', full_name: 'Priya Sharma' },
    actor: { id: DEMO_USER_ID, email: 'admin@tih.co', full_name: 'Admin User' },
  },
  {
    id: 'a2', organization_id: DEMO_ORG_ID, talent_id: 'p4',
    action: 'APPROVED', description: 'Arjun Mehta was approved',
    created_by: DEMO_USER_ID, created_at: '2024-04-30T16:00:00Z',
    talent: { id: 'p4', full_name: 'Arjun Mehta' },
    actor: { id: DEMO_USER_ID, email: 'admin@tih.co', full_name: 'Admin User' },
  },
  {
    id: 'a3', organization_id: DEMO_ORG_ID, talent_id: 'p9',
    action: 'PROFILE_CREATED', description: 'New profile added for Ananya Gupta',
    created_by: DEMO_USER_ID, created_at: '2024-05-12T09:00:00Z',
    talent: { id: 'p9', full_name: 'Ananya Gupta' },
    actor: { id: DEMO_USER_ID, email: 'admin@tih.co', full_name: 'Admin User' },
  },
  {
    id: 'a4', organization_id: DEMO_ORG_ID, talent_id: 'p5',
    action: 'EVALUATION_ADDED', description: 'Evaluation submitted for Divya Nair',
    created_by: DEMO_USER_ID, created_at: '2024-05-20T14:00:00Z',
    talent: { id: 'p5', full_name: 'Divya Nair' },
    actor: { id: DEMO_USER_ID, email: 'admin@tih.co', full_name: 'Admin User' },
  },
  {
    id: 'a5', organization_id: DEMO_ORG_ID, talent_id: 'p2',
    action: 'PROFILE_CREATED', description: 'New profile added for Rahul Verma via Referral',
    created_by: DEMO_USER_ID, created_at: '2024-04-02T09:00:00Z',
    talent: { id: 'p2', full_name: 'Rahul Verma' },
    actor: { id: DEMO_USER_ID, email: 'admin@tih.co', full_name: 'Admin User' },
  },
]

export const demoStats: DashboardStats = {
  total: 10,
  new_count: 2,
  under_review: 2,
  shortlisted: 2,
  approved: 2,
  rejected: 1,
  engaged: 1,
}

export const demoSourceStats = [
  { source: 'LinkedIn', count: 3 },
  { source: 'Referral', count: 2 },
  { source: 'WhatsApp', count: 2 },
  { source: 'Job Portal', count: 1 },
  { source: 'Email', count: 1 },
  { source: 'Manual', count: 1 },
]

export const demoTypeStats = [
  { type: 'Trainer', count: 2 },
  { type: 'Consultant', count: 2 },
  { type: 'Employee', count: 2 },
  { type: 'Freelancer', count: 1 },
  { type: 'Mentor', count: 1 },
  { type: 'Speaker', count: 1 },
  { type: 'Contractor', count: 1 },
]

export const demoPipelineData = [
  { name: 'New', value: 2, fill: '#3b82f6' },
  { name: 'Under Review', value: 2, fill: '#f59e0b' },
  { name: 'Shortlisted', value: 2, fill: '#8b5cf6' },
  { name: 'Approved', value: 2, fill: '#10b981' },
  { name: 'Engaged', value: 1, fill: '#06b6d4' },
  { name: 'Rejected', value: 1, fill: '#ef4444' },
]
